cmd/manager/main.go

```diff
@@ -359,8 +366,11 @@ func startOperator(stopChan <-chan struct{}) error {
 
 	// Create a new Cmd to provide shared dependencies and start components
 	opts := ctrl.Options{
-		Scheme:  clientgoscheme.Scheme,
-		CertDir: viper.GetString(operator.WebhookCertDirFlag),
+		Scheme:                  clientgoscheme.Scheme,
+		CertDir:                 viper.GetString(operator.WebhookCertDirFlag),
+		LeaderElection:          viper.GetBool(operator.EnableLeaderElection),
```

Comment:
I'm trying to think about the problem having 2 operators with leader election enabled might cause.

The major (only?) one I see is that I think there's a possible small time window where 2 instances of the operators may run concurrently:
- the old leader, after it lost the lease but before it shuts down completely (it is supposed to shut down almost immediately but well, you never know which goroutine gets priority) - at this point it may still be creating Pods or interacting with Elasticsearch cluster settings
- the new leader, right after it got the lease

IIUC there's an [extra 5sec](https://github.com/kubernetes-sigs/controller-runtime/blob/2d69b1e99c8bf01ab8dfe9b45474f06e0129894e/pkg/manager/internal.go#L49-L52) to prevent this from happening but I don't think there's any stronger guarantee.

In that small time period, expectations from the 1st operator may not be satisfied in the 2nd one: 1st operator creates a master and 2nd one does not see it yet. As part of startup the 2nd operator (new leader) waits for its cache to be synced before starting to reconcile but we may still hit the race condition described above, I think.

Anyway, it's better to have leader election enabled than not have anything at all.

I'm wondering if this race condition is unlikely enough to happen so we can ignore it, and switch to a Deployment instead of a StatefulSet for the operator.

Reply:
Related issue in operator-sdk: https://github.com/operator-framework/operator-sdk/issues/136
It looks like they implemented an additional election mechanism with different constraints: https://github.com/operator-framework/operator-sdk/pull/1052/files

Reply:
I think I agree with what @barkbay said in https://github.com/elastic/cloud-on-k8s/pull/3632#issuecomment-680764147: perfect leader election is hard and we probably don't need it here. The current mechanism seems good enough for what we want to achieve.